compiler/rustc_infer/src/traits/mod.rs

```diff
@@ -105,6 +109,65 @@ pub enum FulfillmentErrorCode<'tcx> {
     CodeAmbiguity,
 }
 
+pub struct ObligationsDedup<'a, 'tcx, T> {
+    obligations: &'a mut Vec<Obligation<'tcx, T>>,
+}
+
+impl<'a, 'tcx, T: 'tcx> ObligationsDedup<'a, 'tcx, T>
+where
+    T: Clone + Hash + Eq,
+{
+    pub fn from_vec(vec: &'a mut Vec<Obligation<'tcx, T>>) -> Self {
+        ObligationsDedup { obligations: vec }
+    }
+
+    pub fn extend<'b>(&mut self, iter: impl ExactSizeIterator<Item = Cow<'b, Obligation<'tcx, T>>>)
```

Comment:
I'm somewhat confused. Why does `Item` need to be a `Cow`? I'm not seeing where borrows make sense, but I might be missing something ^_^

Wouldn't passing a `&[Obligation<'tcx, T>]` give us maybe better perf? Because doing `obligations.into_iter().map(...)` in the callers is incurring new allocations in the hot path.

Reply:
> I'm not seeing where borrows make sense, but I might be missing something ^_^

Items from the projection cache are borrowed, items from a fresh project_type + normalization are owned.

> Because doing obligations.into_iter().map(...) in the callers is incurring new allocations in the hot path.

Do you mean `obligations.extend(result.obligations.into_iter().map(Cow::Owned));`? That doesn't allocate new obligations, it just stuffs already owned ones into a `Cow::Owned` enum variant.